### ISSUE TITLE
How to access the caching layer while writing custom decoders

### ISSUE THREAD

**Issue by mushtaq** [sirthias/borer#305]
It seems, Borer has an internal caching layer for efficient decoding of `Flat` encoded of ADTs that avoids double-pass decoding. Here is [the relevant section](https://sirthias.github.io/borer/borer-derivation/map-based.html#alternative-encoding-of-adts) in the docs:

> Due the unordered nature of JSON objects/maps it makes efficient single-pass parsing impossible in the general case, since the information, how to interpret an object’s members (the type of the object) might be delayed until the very end of the input file.
This requires either potentially unbounded caching or a second pass over the input. borer’s approach is the former (caching) with a configurable bound on the cache size (triggering an exception, when exceeded).

Is there a way to make use of the caching layer while writing custom decoders? My use case is as follows:

```scala
case class Param[T](valueType: String, value: T)
object Param {
  val decoders = Map(
    "Int"    -> MapBasedCodecs.deriveDecoder[Param[Int]],
    "String" -> MapBasedCodecs.deriveDecoder[Param[String]]
  )

  implicit lazy val paramDec: Decoder[Param[_]] = {
    Decoder[Map[String, Element]].map[Param[_]] { paramAST =>
      val decoder: Decoder[Param[_]] = {
        paramAST("valueType") match {
          case Dom.StringElem(value) => Param.decoders(value)
        }
      }.asInstanceOf[Decoder[Param[_]]]

      val bytes = Json.encode(paramAST).toByteArray
      Json.decode(bytes).to[Param[_]](decoder).value
    }
  }
}

// in the test

"two pass decoding" in {
    // json provided by a third party service follows kind of 'flat' representation
    val json = """
      |[
      | {
      |   "valueType": "Int",
      |   "value": 10
      | },
      | {
      |   "valueType": "String",
      |   "value": "abc"
      | }
      |]
      |""".stripMargin

    import Param.paramDec
    val result = Json.decode(json.getBytes()).to[List[Param[_]]].value
    println(result)
  }
```

I would like to avoid decoding twice in paramDec. 

It is possible to avoid the whole situation by either a) ask the service to send the `nested` json indexed by valueTypes or b) changing the data model, or c) creating new intermediate ones etc. 

But, I want to first find out if there is any way of using the above mentioned caching layer/some internal API to achieve the same. I thought `mapWithReader` will do the job, but I think it is meant for something else.

**Comment by sirthias** [comment]
The internal caching infrastructure is intentionally _not_ a public API, since it's rather low-level as well as very mutable and its proper use requires understanding the inner workings of _borer_ to a much deeper degree than healthy. In other words it's much to easy to shoot yourself in the foot with it.

There might be, however, a relatively easy and fast solution in your special case if the external service always renders the `valueType` member first rather than at random points in the JSON objects you receive. Is that the case?

**Comment by mushtaq** [comment]
> There might be, however, a relatively easy and fast solution in your special case if the external service always renders the `valueType` member first rather than at random points in the JSON objects you receive. Is that the case?

Unfortunately, that is not the case. 

But I understand your reasoning behind keeping the caching API internal. Just wanted to confirm if I have not missed out on something which exists without documentation.

**Comment by sirthias** [comment]
Ok.
Then, maybe there is a way to simply use the existing Flat AdtEncoding for your use case?
In your code snippet above it looks like you are actually rebuilding what is already there, in a less efficient form.

**Comment by sirthias** [comment]
I'm thing about sth along these lines:

```scala
sealed trait Param[T] {
   def value: T
}

object Param {
    @key("Int") final case class IntParam(value: Int) extends Param[Int]
    @key("String") final case class StringParam(value: Int) extends Param[String]
}

implicit val flatAdtEncoding =
  AdtEncodingStrategy.flat(typeMemberName = "valueType")

implicit val paramCodec = deriveAllCodecs[Param[_]]
```

**Comment by mushtaq** [comment]
Yes, that is certainly a clean encoding.

I am still trying to understand why `val bytes = Json.encode(paramAST).toByteArray` is required in my original solution? Without exposing the internal caching layer, is it not be possible to have an API like below?

```scala
Decoder[Map[String, Element]].mapWithInput[Param[_]] { (paramAST, byteArray) =>
  val decoder: Decoder[Param[_]] = {
    paramAST("valueType") match {
      case Dom.StringElem(value) => Param.decoders(value)
    }
  }.asInstanceOf[Decoder[Param[_]]]

  Json.decode(byteArray).to[Param[_]](decoder).value
}
```

In this example, `byteArray` represents the Array[Byte] consumed to create `paramAST`. Maybe the streaming nature of borer does not allow this?

**Comment by sirthias** [comment]
> In this example, byteArray represents the Array[Byte] consumed to create paramAST. Maybe the streaming nature of borer does not allow this?

Yes, exactly. You cannot easily "unread" bytes or re-read what you have already consumed, as this would immediately raise quite a few hard questions.

However, you can test for a match against the next bytes _without_ actually reading them.
This is what the `reader.tryReadXXX` methods are for.
However, this only works on the lowest level and doesn't compose. So, if the `valueType` member isn't the very first, you'd have to iteratively look for it and cache everything that you are skipping, so you can re-read it later.
This is complicated and requires careful buffer management in order to not run into a buffer overflow/DoS vulnerabilities, which is why _borer_ implements it itself via the `AdtEncodingStrategy.flat`, so you don't have to.

So, in short, simply try to use the `AdtEncodingStrategy.flat`, which appears to be doing exactly what you need, no?

**Comment by mushtaq** [comment]
> This is complicated and requires careful buffer management in order to not run into a buffer overflow/DoS vulnerabilities, which is why _borer_ implements it itself via the `AdtEncodingStrategy.flat`, so you don't have to.
> 
> So, in short, simply try to use the `AdtEncodingStrategy.flat`, which appears to be doing exactly what you need, no?

Thanks for explaining this! And yes, `AdtEncodingStrategy.flat` based solution looks like the best approach here.

**Comment by mushtaq** [comment]
> There might be, however, a relatively easy and fast solution in your special case if the external service always renders the `valueType` member first rather than at random points in the JSON objects you receive. Is that the case?

@sirthias I also want to fully understand this option. 

Your subsequent comment indicates that reader.tryReadString("valueType") will be used which will return a Boolean.

But what after that? Not able to visualise the full solution here.

**Comment by sirthias** [comment]
After you've read the String `"valueType"` you know that the next element is going to be the `valueType` value. So you read that one and branch out to whatever logic is required for reading the specific value contents.
The benefit is that you know what structure to deserialize _before_ reading it, which means you don't have to cache it.

**Comment by mushtaq** [comment]
Ok, so instead of `reader.tryReadString("valueType")`, I will have to say `val valueType = r.readString("valueType").readString()`, and then branch out based on `valueType`.

Maybe `tryReadString` will just help in throwing a custom exception if it returns `false`.

**Comment by sirthias** [comment]
Well, you would likely have sth like this:

```scala
if (r.tryReadString("valueType")) {
  r.readString match {
     case "Dog" => readDog(r)
     case "Cat" => readCat(r)
     ...
  }
} else {
  ... // handle valueType not being the first member
}
```

The problem with this approach is that `readDog` can't be a simple `dogDecoder.read(r)`, because a complete decoder would require the first element to be an `ArrayStart/Header` or `MapStart/Header` element.
But in this case this opening element has already been consumed and the cursor is now _inside_ of the map or array.

**Comment by mushtaq** [comment]
Ok, I see. I tried it so understand what you mean by 'cursor is now inside of the map'. Thanks!

**Comment by mushtaq** [comment]
With the latest version supporting transcoding, I realized that these 2 steps from my original decoder:

```scala
val bytes = Json.encode(paramAST).toByteArray
Json
  .decode(bytes)
  .to[Param[_]](decoder)
  .value
```

Can be replaced with a single step:
```scala
Json
  .transEncode(paramAST)
  .transDecode
  .to[Param[_]](decoder)
  .value
```

This will have the benefit of single pass decoding I was looking for. Did I get it right, @sirthias

**Comment by sirthias** [comment]
Yes, @mushtaq, the second snippet you show produces identical output to the first one but is much more efficient, since no actual intermediate byte array representation has to be generated.
👍

**Comment by mushtaq** [comment]
Thanks for confirming!
